alpa/create_state_parallel.py

```diff
@@ -0,0 +1,176 @@
+                                                  spec.sharding_specs[1:]):
+                    indices = pxla.spec_to_indices(array.shape, sharding_spec)
+                    dis_array = self.mesh_group[mesh_id].shard_args_to_arrays(
+                        (array.aval,), (indices,), (sharding_spec,),
+                        (array,))[0]
+                    distributed_arrays.append(dis_array)
+                outputs[idx] = ReplicatedDistributedArray(
+                    meshes, distributed_arrays)
+
+        return outputs
+
+
+def compile_create_state_executable(fun, in_tree, out_tree_thunk,
+                                    static_argnums, donated_invars, train_step,
+                                    other_args, *avals):
+    # Trace to get jaxpr and HloModule
+    jaxpr, out_avals, consts = pe.trace_to_jaxpr_final(fun, avals)
+    closed_jaxpr = ClosedJaxpr(jaxpr, consts)
+    out_tree = out_tree_thunk()
+    state_aval = tree_unflatten(out_tree, out_avals)
+
+    name = f"{fun.__name__}_create_state_parallel"
+    backend = xb.get_backend("gpu")
+    hlo_module = jaxpr_to_hlo_module(name, closed_jaxpr, donated_invars,
+                                     backend)
+
+    # Compile train_step to get the placement specs.
+    executable = train_step.get_executable(state_aval, other_args)
+    placement_specs = executable.get_placement_specs()[0]
+    placement_specs, _ = tree_flatten(placement_specs)
+
+    if isinstance(executable,
+                  (NormalMeshDriverExecutable, GradAccMeshDriverExecutable)):
+        sharding_protos = []
+        for spec in placement_specs:
+            assert len(spec.mesh_ids) == 1
+            sharding_protos.append(spec.sharding_specs[0].sharding_proto())
+
+        physical_mesh = executable.physical_mesh
+
+        # Run sharding propagation
+        xe.set_hlo_module_output_shardings(hlo_module, sharding_protos)
+        hlo_module, strategy_config = run_auto_sharding_pass(
+            hlo_module, avals, out_avals, donated_invars,
+            physical_mesh.get_logical_mesh(
+                executable.strategy_config.logical_mesh_shape), "single", 1,
+            AutoShardingOption(enable_auto_sharding=False))
+
+        return NormalMeshDriverExecutable(physical_mesh, hlo_module,
+                                          strategy_config, avals, out_avals,
+                                          [False] * len(avals))
+    else:
+        # Construct a new pipelined jaxpr
+        outvars = jaxpr.outvars
+
+        var2mesh = {}  # Dict[var -> mesh_id]
+        eqn2mesh = {}  # Dict[eqn_idx -> mesh_id]
+
+        output_shardings = []
+        for var, spec in zip(outvars, placement_specs):
+            if isinstance(var, Var):
+                var2mesh[var] = spec.mesh_ids[0]
+            output_shardings.append(spec.sharding_specs[0])
+
+        num_meshes = len(executable.mesh_group)
+
+        propagate_mesh_assignment(jaxpr, var2mesh, eqn2mesh)
```

Comment:
Apply Gradient in Pipeshard parallel's compile executable has a likewise function. I'll try to merge the two and move to utils.py

Reply:
They have a slight difference:
- in apply_grad, we can replicate the equations to favor recomputation over communication. 
- in create_state, due to our use of stateful random number generation, replicated computation can generate different results on different device meshes. So we cannot replicate equations. We have to use send/recv to send the temp results among meshes. To avoid the deadlock, we also need to fix an order.